src/sentry/spans/buffer_v2.py

```diff
@@ -0,0 +1,290 @@
+3. Add the ingested span's payload to the set under `set_key`.
+4. To a "global queue", we write the set's key, sorted by timeout.
+
+Eventually, flushing cronjob looks at that global queue, and removes all timed
+out keys from it. Then fetches the sets associated with those keys, and deletes
+the sets.
+
+This happens in two steps: Get the to-be-flushed segments in `flush_segments`,
+then the consumer produces them, then they are deleted from Redis
+(`done_flush_segments`)
+
+On top of this, the global queue is sharded by partition, meaning that each
+consumer reads and writes to shards that correspond to its own assigned
+partitions. This means that extra care needs to be taken when recreating topics
+or using spillover topics, especially when their new partition count is lower
+than the original topic.
+"""
+
+from __future__ import annotations
+
+from collections.abc import Sequence
+from typing import NamedTuple
+
+import rapidjson
+from django.conf import settings
+from sentry_redis_tools.clients import RedisCluster, StrictRedis
+
+from sentry.utils import metrics, redis
+
+# This SegmentId is an internal identifier used by the redis buffer that is
+# also directly used as raw redis key. the format is
+# "span-buf:s:{project_id:trace_id}:span_id", and the type is bytes because our redis
+# client is bytes.
+#
+# The segment ID in the Kafka protocol is actually only the span ID.
+SegmentId = bytes
+
+
+def get_redis_client() -> RedisCluster[bytes] | StrictRedis[bytes]:
+    return redis.redis_clusters.get_binary(settings.SENTRY_SPAN_BUFFER_CLUSTER)
+
+
+add_buffer_script = redis.load_redis_script("spans/add-buffer.lua")
+
+
+def segment_to_span_id(segment_id: SegmentId) -> bytes:
+    return segment_id.rsplit(b":", 1)[1]
+
+
+# fun fact: namedtuples are faster to construct than dataclasses
+class Span(NamedTuple):
+    trace_id: str
+    span_id: str
+    parent_span_id: str | None
+    project_id: int
+    payload: bytes
+    is_segment_span: bool = False
+
+
+class RedisSpansBufferV2:
+    def __init__(
+        self,
+        assigned_shards: list[int],
+        span_buffer_timeout_secs: int = 60,
+        span_buffer_root_timeout_secs: int = 10,
+        redis_ttl: int = 3600,
+    ):
+        self.client: RedisCluster[bytes] | StrictRedis[bytes] = get_redis_client()
+        self.assigned_shards = list(assigned_shards)
+        self.span_buffer_timeout_secs = span_buffer_timeout_secs
+        self.span_buffer_root_timeout_secs = span_buffer_root_timeout_secs
+        self.redis_ttl = redis_ttl
+
+    # make it pickleable
+    def __reduce__(self):
+        return (
+            RedisSpansBufferV2,
+            (
+                self.assigned_shards,
+                self.span_buffer_timeout_secs,
+                self.span_buffer_root_timeout_secs,
+                self.redis_ttl,
+            ),
+        )
+
+    def _is_root_span(self, span: Span) -> bool:
+        return span.parent_span_id is None or span.is_segment_span
```

Comment:
`is_segment` considers this already. However, we need to add a check somewhere if the parent span is in a different project. I'm fine if we leave this for a follow-up, but let's record this as a TODO.